extmod/asyncio/stream.py

```diff
@@ -170,6 +171,47 @@ async def start_server(cb, host, port, backlog=5):
     return srv
 
 
+# Helper task to run a TCP stream server.
+# Callbacks (i.e. connection handlers) may run in a different taskgroup.
+async def run_server(cb, host, port, backlog=5, taskgroup=None):
```

Comment:
cpython doesn't have a taskgroup-aware way to run a server, it simply starts a new task for each client (for some value of "simply", its code is rather complex for historical reasons related to their protocol/transport split – in hindsight, that was a rather bad design decision).

That process doesn't play well with taskgroups. A taskgroup *guarantees* that when you leave it, all of its resources, subtasks and whatnot are done with. As soon as you start a non-taskgrouped task from within that, all those nice guarantees go out the window. You can't work around this problem without introducing a race condition. 

Yes this is a micropython extension and thus something that we ordinarily don't want to do, but CPython simply doesn't have an equivalent.

Reply:
… which reminds me that the whole taskgroup thing needs documentation. Duh. Working on it.

Reply:
So on cpython, if you run start _server in a TaskGroup it'll run, but not get cleaned up if the TaskGroup exits?

Do you know if this has raised as an issue for cpython?

Reply:
I have no idea. I plan to investigate that (and submit a PR if necessary).

The "problem" is that most likely the issue hasn't come up yet because people who use taskgroups tend to do that with anyio or even trio. Both have their own wrappers to address this problem, which unfortunately are not a good fit for micropython.

Reply:
No I didn't yet, no free time to shepherd a feature like that, esp. since taskgroups still are an under-utilized (and poorly understood, among old-time asyncio adherents) feature.

I'll drop it for now and move the code to `examples/` until we find a better home for it.